src/packet_line.rs

```diff
@@ -27,25 +32,37 @@ impl PktLine<'_> {
     pub fn encode_to(&self, buf: &mut BytesMut) -> Result<(), Error> {
         match self {
             Self::Data(data) => {
-                write!(buf, "{:04x}", data.len() + 4)?;
-                buf.extend_from_slice(data);
+                for chunk in data.chunks(MAX_DATA_LEN) {
+                    write!(buf, "{:04x}", chunk.len() + 4)?;
+                    buf.extend_from_slice(chunk);
+                }
             }
             Self::SidebandData(packfile) => {
                 // split the buf off so the cost of counting the bytes to put in the
                 // data line prefix is just the cost of `unsplit` (an atomic decrement)
                 let mut data_buf = buf.split_off(buf.len());
 
-                data_buf.put_u8(1); // sideband, 1 = data
                 packfile.encode_to(&mut data_buf)?;
 
                 // write into the buf not the data buf so it's at the start of the msg
-                write!(buf, "{:04x}", data_buf.len() + 4)?;
-                buf.unsplit(data_buf);
+                if data_buf.len() + 5 <= MAX_DATA_LEN - 1 {
+                    write!(buf, "{:04x}", data_buf.len() + 5)?;
+                    buf.put_u8(1); // sideband, 1 = data
+                    buf.unsplit(data_buf);
+                } else {
+                    for chunk in data_buf.chunks(MAX_DATA_LEN - 1) {
+                        write!(buf, "{:04x}", chunk.len() + 5)?;
+                        buf.put_u8(1); // sideband, 1 = data
+                        buf.extend_from_slice(chunk);
+                    }
+                }
             }
```

Comment:
I'm not sure if the unsplit optimisation actually works any more, maybe could simplify this to just always use chunks?

Reply:
Will look into this over the next couple of days but will merge this for now and bump gcs, thanks for the quick fix!